Source/Scene/ModelExperimental/MorphTargetsPipelineStage.js

```diff
@@ -0,0 +1,255 @@
+import AttributeType from "../AttributeType.js";
+import combine from "../../Core/combine.js";
+import defined from "../../Core/defined.js";
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+import MorphTargetsStageVS from "../../Shaders/ModelExperimental/MorphTargetsStageVS.js";
+import VertexAttributeSemantic from "../VertexAttributeSemantic.js";
+
+/**
+ * The morph targets pipeline stage processes the morph targets and weights of a primitive.
+ *
+ * @namespace MorphTargetsPipelineStage
+ *
+ * @private
+ */
+
+const MorphTargetsPipelineStage = {};
+MorphTargetsPipelineStage.name = "MorphTargetsPipelineStage"; // Helps with debugging
+MorphTargetsPipelineStage.FUNCTION_ID_GET_MORPHED_POSITION =
+  "getMorphedPosition";
+MorphTargetsPipelineStage.FUNCTION_SIGNATURE_GET_MORPHED_POSITION =
+  "vec3 getMorphedPosition(in vec3 position)";
+MorphTargetsPipelineStage.FUNCTION_ID_GET_MORPHED_NORMAL = "getMorphedNormal";
+MorphTargetsPipelineStage.FUNCTION_SIGNATURE_GET_MORPHED_NORMAL =
+  "vec3 getMorphedNormal(in vec3 normal)";
+MorphTargetsPipelineStage.FUNCTION_ID_GET_MORPHED_TANGENT = "getMorphedTangent";
+MorphTargetsPipelineStage.FUNCTION_SIGNATURE_GET_MORPHED_TANGENT =
+  "vec3 getMorphedTangent(in vec3 tangent)";
+
+/**
+ * This pipeline stage processes the morph targets and weights of a primitive,
+ * adding the relevant attribute declarations and functions to the shaders.
+ *
+ * Processes a primitive. This stage modifies the following parts of the render resources:
+ * <ul>
+ *  <li> adds attribute declarations for the morph targets in the vertex shader
+ *  <li> adds the uniform declaration for the morph weights in the vertex shader
+ *  <li> adds functions to apply the morphs in the vertex shader
+ * </ul>
+ *
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} primitive The primitive.
+ *
+ * @private
+ */
+MorphTargetsPipelineStage.process = function (renderResources, primitive) {
+  const shaderBuilder = renderResources.shaderBuilder;
+
+  shaderBuilder.addDefine(
+    "HAS_MORPH_TARGETS",
+    undefined,
+    ShaderDestination.VERTEX
+  );
+
+  addGetMorphedAttributeFunctionDeclarations(shaderBuilder);
+
+  for (let i = 0; i < primitive.morphTargets.length; i++) {
```

Comment:
Hm, I see why this is a bit tricky to refactor, the morph targets are stored by index, then by semantic. However the shader code cares about the semantic first, then the index. Is there a reasonable way to reorganize this?